### PR TITLE
define-array and using the array index implicitly

### PR DESCRIPTION
**Pull request by DrewMartin**
Implementation of https://github.com/Shopify/twine/issues/73

This lets you use `define-array` (or `data-define-array`) similar to define, except it stores the returned values in an array and will push new values to the end. We also store some metadata about array indexes behind the scenes to know which elements correspond to which iterations of an array. Every element (that has bindings) have the same array indexes of its parent, with any new indexes as well. When a new context is entered the nodes inside of it start with a fresh set of array indexes as well, unable to see the array indexes on nodes outside of the current context.
For example:
```html
<div data-define-array="{key: {value: 5, foo: function(){ return 12; }}}">
  <div data-define-array="{nested: {value: 50}}">
    <span bind="key.value">5</span> <!-- key[0].value -->
    <span bind="key.value">12</span> <!-- key[0].foo() -->
    <span bind="nested.value">50</span> <!-- nested[0].value -->
  </div>
  <div data-define-array="{nested: {value: 75}}">
    <span bind="key.value">5</span> <!-- key[0].value -->
    <span bind="key.value">12</span> <!-- key[0].foo() -->
    <span bind="nested.value">75</span> <!-- nested[1].value -->
  </div>
</div>
<div data-define-array="{key: {value: 10, foo: function(){ return 99; }}}">
  <div data-define-array="{nested: {value: 100}}">
    <span bind="key.value">10</span> <!-- key[1].value -->
    <span bind="key.value">99</span> <!-- key[1].foo() -->
    <span bind="nested.value">100</span> <!-- nested[2].value -->
  </div>
  <div data-define-array="{nested: {value: 125}}">
    <span bind="key.value">10</span> <!-- key[1].value -->
    <span bind="key.value">99</span> <!-- key[1].foo() -->
    <span bind="nested.value">125</span> <!-- nested[3].value -->
  </div>
</div>
<div data-define-array="{key: {value: 42}}" context="key"> <!-- key[2] -->
  <!-- has no indexes available -->
  <span bind="value">42</span>
</div>
```

This takes two main changes:
* when building the keypath for a binding, which is used for getting and setting simple values, where it would be `["nested", "value"]` for `nested.value`, instead it becomes `["nested", 0, "value"]` (or whatever the correct array index is, pulled from the node itself). This will only happen for the first key in the keypath because that's the potential array on the context and everything else is nested within the context and not created with `define-array`, at least not on the element being evaluated. (you could have a `define-array` -> `context` -> `define-array` but the outer node wouldn't have access to the array indexes on the inner `define-array`)
* when the binding does not resolve to a keypath, so basically when it's a function call, we create a hash of objects inside the array(s) and use it as the innermost `with` expression, inside of the `with($context)`, so that these take precedence. Effectively it does something like this:
```js
with({key: [1, 2, 3]}) {
  with({key: 1}) { // Or {key: 2}, or {key: 3} depending on what the index is for this node in the array `key` 
    return // Code here
  }
}
```

@pushrax @Shopify/tnt

### CODE REVIEW

File: lib/assets/javascripts/twine.coffee [hunk 1]
```diff
@@ -147,6 +166,13 @@
         return context
       node = node.parentNode if child
 
+  getIndexesForElement = (node) ->
+    firstContext = null
+    while node
+      return elements[id]?.indexes if id = node.bindingId
```
Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: Since we assign the same indexes to all child nodes that have any bindings, if a node has a bindingId it will also have the set of indexes to use, if any.

File: lib/assets/javascripts/twine.coffee [hunk 2]
```diff
@@ -46,33 +46,47 @@
...
 
-  bind = (context, node, forceSaveContext) ->
+  bind = (context, node, indexes, forceSaveContext) ->
     currentBindingCallbacks = []
     if node.bindingId
       Twine.unbind(node)
 
+    if defineArrayAttr = Twine.getAttribute(node, 'define-array')
+      newIndexes = defineArray(node, context, defineArrayAttr)
+      for key, value of indexes || {} when !newIndexes[key]?
```
Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: Didn't know you could use `when` like that 👍

File: lib/assets/javascripts/twine.coffee [hunk 3]
```diff
@@ -46,33 +46,47 @@
...
+      indexes = newIndexes
+      # register the element early because subsequent bindings on the same node might need to make use of the index
+      element = findOrCreateElementForNode(node)
+      element.indexes = indexes
+
     for type, binding of Twine.bindingTypes when definition = Twine.getAttribute(node, type)
-      element = {bindings: []} unless element  # Defer allocation to prevent GC pressure
+      element = findOrCreateElementForNode(node)
+      element.bindings ?= []
+      element.indexes ?= indexes
```
Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: You're sure you can use `?=` here? That checks explicitly for `null` and nothing else. If `element.indexes` is `undefined`, it will stay `undefined` in this case. Is it maybe safer to use `element.indexes ||= indexes` and cover both cases?

File: lib/assets/javascripts/twine.coffee [hunk 4]
```diff
@@ -147,6 +166,13 @@
         return context
       node = node.parentNode if child
 
+  getIndexesForElement = (node) ->
+    firstContext = null
+    while node
+      return elements[id]?.indexes if id = node.bindingId
+      node = node.parentNode
+    return
```
Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: No need for this.

File: lib/assets/javascripts/twine.coffee [hunk 5]
```diff
@@ -219,6 +259,20 @@
...
   requiresRegistry = (args) -> /\$registry/.test(args)
 
+  nodeHasArrayIndexes = (node) ->
+    return unless node.bindingId?
+    !!elements[node.bindingId]?.indexes?
+
+  arrayPointersForNode = (node, context) ->
+    return {} unless node.bindingId?
+    indexes = elements[node.bindingId]?.indexes
+    return {} unless !!indexes?
```
Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: This is clever, but a pretty unclear. What do you think of making the "allow 0" aspect a little more explicit?

`return {} unless indexes || indexes == 0`

**Comment by dan-menard** [comment]
A few code comments, but the syntax and overall implementation look solid to me 👍

File: lib/assets/javascripts/twine.coffee (hunk 5 above)
Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: I don't know why I have !!, `indexes?` alone would do it

File: lib/assets/javascripts/twine.coffee (hunk 3 above)
Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: It checks for `== null` in javascript, and `undefined == null` is true. The `?` operator in cs checks if something is undefined or null

Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: Ah! Yeah, sorry, mistook the == for === when I ran that through a to-JS converter. Code's good as-is 👍

File: lib/assets/javascripts/twine.coffee (hunk 5 above)
Comment (dan-menard) [inline lib/assets/javascripts/twine.coffee]: Oh, true! Yeah, I'm fine with just `indexes?`

File: dist/twine.js [hunk 6]
```diff
@@ -61,55 +61,84 @@
...
       }
-      ref = Twine.bindingTypes;
-      for (type in ref) {
-        binding = ref[type];
+      if (defineArrayAttr = Twine.getAttribute(node, 'define-array')) {
+        newIndexes = defineArray(node, context, defineArrayAttr);
+        ref = indexes || {};
+        for (key in ref) {
+          value = ref[key];
+          if (newIndexes[key] == null) {
```
Comment (lemonmade) [inline dist/twine.js]: `hasOwnProperty` is probably more what you are after here.

File: lib/assets/javascripts/twine.coffee (hunk 2 above)
Comment (lemonmade) [inline lib/assets/javascripts/twine.coffee]: Can you pull the `indexes || {}` part up to a dedicated variable? This line is really hard to parse. Also, `hasOwnProperty` is probably more what you are after for the `newIndexes[key]` part.

Comment (qq99) [inline lib/assets/javascripts/twine.coffee]: `indexes = {}` default in the method signature?

File: lib/assets/javascripts/twine.coffee [hunk 7]
```diff
@@ -46,33 +46,47 @@
...
+      newIndexes = defineArray(node, context, defineArrayAttr)
+      for key, value of indexes || {} when !newIndexes[key]?
+        newIndexes[key] = value
+      indexes = newIndexes
+      # register the element early because subsequent bindings on the same node might need to make use of the index
+      element = findOrCreateElementForNode(node)
+      element.indexes = indexes
+
     for type, binding of Twine.bindingTypes when definition = Twine.getAttribute(node, type)
-      element = {bindings: []} unless element  # Defer allocation to prevent GC pressure
```
Comment (qq99) [inline lib/assets/javascripts/twine.coffee]: how did this work? I don't get the `unless element` on same line

Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: Since this is in a loop the first iteration will initialize it. It looks like it wasn't being initialized outside because of the large amount of elements in the dom that wouldn't have any bindings, so that we didn't create and then throw away the `element` object, making the GC work harder

File: lib/assets/javascripts/twine.coffee (hunk 2 above)
Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: I don't want to do indexes as a default arg because that would mean a lot of elements storing an empty object as the indexes, which means we have to check the number of keys in an object to tell if an element has indexes rather than just a null check.

Better to keep it as null/undefined when there are no indexes to store.

Comment (DrewMartin) [inline lib/assets/javascripts/twine.coffee]: (It would also affect the assignment on line 89)

**Comment by DrewMartin** [comment]
Everyone cool with this? 🚢 ?

**Comment by tsov** [comment]
👍  we can iterate on it too.

Would you mind adding a commit with README docs for `define-array` to https://github.com/Shopify/twine/pull/77 ?

**Comment by DrewMartin** [comment]
It doesn't look like it really belongs in the readme so much as http://shopify.github.io/twine/ or the docs @dan-menard's working on, but I'll get it added to one or both
